### PR TITLE
Update expectation for projection of slice reference.

### PR DESCRIPTION
**Pull request by celinval**
### Description of changes: 

Our expectation did not considered the array decay when creating fat pointers. This change relaxes our expectations to account for array decay and adds more tests.

### Resolved issues:

This fixes some warnings from #277 but not all.

### Testing:

* How is this change tested?

* Is this a refactor change?

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: src/kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs [hunk 1]
```diff
@@ -461,9 +477,10 @@ impl<'tcx> GotocCtx<'tcx> {
     /// If it passes through a fat pointer along the way, it stores info about it,
     /// which can be useful in reconstructing fat pointer operations.
     pub fn codegen_place(&mut self, p: &Place<'tcx>) -> ProjectedPlace<'tcx> {
-        debug!("codegen_place: {:?}", p);
+        debug!(place=?p, "codegen_place");
```
Comment (danielsn) [inline src/kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs:480]: Nice

File: src/kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs [hunk 2]
```diff
@@ -754,7 +757,7 @@ impl<'tcx> GotocCtx<'tcx> {
         if self.use_slice_fat_pointer(mir_type) {
             let pointer_name = match mir_type.kind() {
                 ty::Slice(..) => self.ty_mangled_name(mir_type),
-                ty::Str => "str".intern(),
+                ty::Str => "refstr".intern(),
```
Comment (danielsn) [inline src/kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs:760]: I wonder if there is a better name we could use here.

File: tests/expected/raw_slice/slice.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+//! This test case has a bunch of checks related to structures using raw slices ([T]).
+use std::mem;
+
+/// Non-sized structure with a sized element and an unsized element.
+struct NonEmptySlice {
+    first: u8,
+    others: [u8],
+}
+
+/// Non-sized structure with a sized element and an unsized element.
```
Comment (danielsn) [inline tests/expected/raw_slice/slice.rs:14]: Only an un sized element

File: tests/expected/raw_slice/slice.rs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    ///
+    /// I.e.: The length is only relative to the unsized part of the structure.
+    /// The total size of the object is the `size_of(sized_members) + (unsize_len * size_of::<u8>())`
+    ///
+    fn new(bytes: &mut [u8]) -> &Self {
+        assert!(bytes.len() > 0, "This requires at least one element");
+        let unsized_len = bytes.len() - 1;
+        unsafe {
+            let ptr = std::ptr::slice_from_raw_parts_mut(bytes.as_mut_ptr(), unsized_len);
+            unsafe { &*(ptr as *mut Self) }
```
Comment (danielsn) [inline tests/expected/raw_slice/slice.rs:39]: Can we assert anything here

File: tests/expected/raw_slice/slice.rs [hunk 5]
```diff
@@ -0,0 +1,102 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: --unwind 3
+
+//! This test case has a bunch of checks related to structures using raw slices ([T]).
+use std::mem;
+
+/// Non-sized structure with a sized element and an unsized element.
+struct NonEmptySlice {
```
Comment (danielsn) [inline tests/expected/raw_slice/slice.rs:9]: I don't think we need it, but should we see what happens with different repr here?

File: src/kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs [hunk 6]
```diff
@@ -320,7 +336,7 @@ impl<'tcx> GotocCtx<'tcx> {
                         inner_goto_expr.member("data", &self.symbol_table)
                     }
                     ty::Adt(..) if self.is_unsized(inner_mir_typ) => {
-                        // in cbmc-reg/Strings/os_str_reduced.rs, we see
+                        // in tests/kani/Strings/os_str_reduced.rs, we see
```
Comment (danielsn) [inline src/kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs:339]: Good catch

**Comment by danielsn** [comment]
Nice work. Thanks for fixing this!

File: src/kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs (hunk 2 above)
Comment (celinval) [reply src/kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs:760]: Great question. Any suggestion?

File: tests/expected/raw_slice/slice.rs (hunk 5 above)
Comment (celinval) [reply tests/expected/raw_slice/slice.rs:9]: Do you mean add a harness that checks the representation?

File: tests/expected/raw_slice/slice.rs (hunk 3 above)
Comment (celinval) [reply tests/expected/raw_slice/slice.rs:14]: Oopps. Copy and paste. LoL

File: tests/expected/raw_slice/slice.rs (hunk 4 above)
Comment (celinval) [reply tests/expected/raw_slice/slice.rs:39]: Can you think about anything specific?

**Review by zhassan-aws — APPROVED** [review]
I haven't looked at the added tests thoroughly, but the code changes look good to me. Thanks for fixing this!

**Pull request merged**